### PR TITLE
support for fields, some more documentation, getting inputs from context, more openapi support with info, support for descriptions, json error responses in context, csrf middleware, fixed makefile, bug fix on pr#14 middlewares + improvement

### PR DESCRIPTION
**Pull request by tiredkangaroo**
list of changes:
- fixed makefile so it properly runs the example code
- **added support for descriptions and fields (still in the way!)** note: formdata not yet supported
- getting inputs from context (ex. body, cookies, headers, etc.)
- json error responses in context (just a small tool)
- **wrote documentation (still in the way!)**
- **support for more information in the openapi schema**
- csrf middleware (still in the way!)
- bugfix #14 -> middlewares on parent router should be present in children + added interface to attach middlewares on specific routers + middleware wrapping evaluation is no longer done at runtime

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -18,6 +18,10 @@ type Config struct {
 	DocsURL string
 	// DocsReload, if true, enables automatic reload on the Swagger documentation page.
 	DocsReload bool
+	// TLSPublicKeyFile specifies the file for the TLS public key (usually .pem or .crt).
+	TLSPublicKeyFile string
+	// TLSPrivateKeyFile specifies the filre for the TLS private key (usually .key).
```
Comment (nikumar1206) [inline app.go]: spelling

File: app.go [hunk 2]
```diff
@@ -16,6 +16,10 @@ type Config struct {
 	DocsURL string
 	// DocsReload, if true, enables automatic reload on the Swagger documentation page.
 	DocsReload bool
+	// TLSPublicKeyFile specifies the file for the TLS public key (usually .pem or .crt).
```
Comment (nikumar1206) [inline app.go]: Should be called Public Cert / Private Key
anything else is incorrect.

File: app.go [hunk 3]
```diff
@@ -73,22 +77,22 @@ func (a *PuffApp) addOpenAPIRoutes() {
 		Name:   "OpenAPI Documentation Router",
 	}
 
-	docsRouter.Get(".json", Field{}, func(c *Context) {
+	docsRouter.Get(".json", "Provides JSON OpenAPI Schema.", nil, func(c *Context) {
```
Comment (nikumar1206) [inline app.go:53]: very much dont like the descriptions.
lets brainstorm a way on fixing them.

File: app.go [hunk 4]
```diff
@@ -114,47 +118,65 @@ func (a *PuffApp) addOpenAPIRoutes() {
...
+	for _, router := range router.Routers {
+		attachMiddlewares((middleware_combo), router)
+	}
+}
 func (a *PuffApp) patchAllRoutes() {
 	a.RootRouter.patchRoutes()
 	for _, r := range a.RootRouter.Routers {
 		r.patchRoutes()
 	}
+	attachMiddlewares(&[]Middleware{}, a.RootRouter)
```
Comment (nikumar1206) [inline app.go:114]: trusting this has been tested.

File: context.go [hunk 5]
```diff
@@ -49,6 +50,36 @@ func (ctx *Context) SetHeader(k, v string) {
 	ctx.ResponseWriter.Header().Set(k, v)
 }
 
+// GetBody returns the request body.
+func (ctx *Context) GetBody() ([]byte, error) {
```
Comment (nikumar1206) [inline context.go:54]: body can only be read once right. should we keep track of if the body has already been read?

Comment (nikumar1206) [reply context.go:54]: im guessing context is a good place to store that?

File: docs/DOCS.md [hunk 6]
```diff
@@ -1,4 +1,6 @@
-# Creating a Puff App
+<h1 style="text-align:center;"> Documentation for Puff </h1>
```
Comment (nikumar1206) [inline docs/DOCS.md:1]: no html, just use #
# Puff
thats all

File: middleware/csrf.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/nikumar1206/puff"
+)
+
+type CSRFMiddlewareConfig struct {
```
Comment (nikumar1206) [inline middleware/csrf.go:10]: this is not how to document the struct. Here is example
// Regexp is the representation of a compiled regular expression.
// A Regexp is safe for concurrent use by multiple goroutines,
// except for configuration methods, such as Longest.
type Regexp struct {
    ...
}
the comments must be placed above the attribute as well.

File: middleware/csrf.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+		}
+	}
+}
+
+// CSRF middleware
+func CSRF() puff.Middleware {
+	return createCSRFMiddleware(DefaultCSRFMiddleware)
+}
+
+func CSRFWithConfig(config *CSRFMiddlewareConfig) puff.Middleware {
```
Comment (nikumar1206) [inline middleware/csrf.go:72]: missing doc

File: middleware/panic.go [hunk 9]
```diff
@@ -9,13 +9,13 @@ import (
 )
 
 type PanicConfig struct {
```
Comment (nikumar1206) [inline middleware/panic.go:12]: docstring please
on this file, since ur already here

File: middleware/tracing.go [hunk 10]
```diff
@@ -6,8 +6,8 @@ import (
 )
 
 type TracingConfig struct {
-	TracerName  string
-	IDGenerator func() string
+	TracerName  string        //TracerName is the name of the response header where the request id will be.
```
Comment (nikumar1206) [inline middleware/tracing.go]: comments go above the attribute

also struct needs docstring, entire file does really.

File: openapi.go [hunk 11]
```diff
@@ -27,6 +30,20 @@ type OpenAPI struct {
...
+// // Definitions contains schemas that can be referenced throughout
+// // the rest of the document. Reference:
+// // https://spec.openapis.org/oas/v3.1.0#schema-object
+// type Definition struct {
+// 	Type       string   `json:"type"`
+// 	Required   []string `json:"required"`
+// 	Properties map[string]Property
+// }
+
+type Property struct {
```
Comment (nikumar1206) [inline openapi.go:44]: docstring

File: response.go [hunk 12]
```diff
@@ -106,7 +106,7 @@ func (f *FileResponse) Handler() func(*Context) {
 // StreamingResponse represents a response that streams content.
 type StreamingResponse struct {
 	StatusCode    int
-	StreamHandler func(*chan ServerSideEvent)
+	StreamHandler func(*chan ServerSideEvent) // Write to this channel to write to the response. Close the channel once you are done.
```
Comment (nikumar1206) [inline response.go]: comments above the attribute.
the line above

File: response.go [hunk 13]
```diff
@@ -170,7 +170,7 @@ func (s StreamingResponse) Handler() func(*Context) {
 
 // RedirectResponse represents a response that sends a redirect to the client.
 type RedirectResponse struct {
-	// StatusCode provides the 3xx status code of the redirect response.
+	// StatusCode provides the 3xx status code of the redirect response. Default: 308.
```
Comment (nikumar1206) [inline response.go:176]: 301 or 308. check please

File: router.go [hunk 14]
```diff
@@ -150,10 +156,16 @@ func (r *Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		}
 		isMatch := route.regexp.MatchString(req.URL.Path)
 		if isMatch && req.Method == route.Protocol {
-			// err := route.Fields.ValidateIncomingAttribute(Field.Responses, "cheese")
+			// err := route.anys.ValidateIncomingAttribute(any.Responses, "cheese")
```
Comment (nikumar1206) [inline router.go]: remove all the comments in this file.

**Review by nikumar1206 — CHANGES_REQUESTED** [review]

File: app.go (hunk 2 above)
Comment (tiredkangaroo) [reply app.go]: resolved

File: app.go (hunk 1 above)
Comment (tiredkangaroo) [reply app.go]: not sure where it says that, im on the latest commit on this branch and it doesn't say "filre" and i didn't make any changes to correct it neither.

Comment (tiredkangaroo) [reply app.go]: but resolved anyway i guess

File: context.go (hunk 5 above)
Comment (tiredkangaroo) [reply context.go:54]: store the body so it can be read multiple times or only keep track if it has already been read?

File: app.go (hunk 1 above)
Comment (tiredkangaroo) [reply app.go]: oh wait it says the file you commented on was Outdated

File: app.go (hunk 4 above)
Comment (tiredkangaroo) [reply app.go:114]: you are right to trust, it has 😁

File: docs/DOCS.md (hunk 6 above)
Comment (tiredkangaroo) [reply docs/DOCS.md:1]: https://stackoverflow.com/questions/14051715/markdown-native-text-alignment

File: middleware/tracing.go (hunk 10 above)
Comment (tiredkangaroo) [reply middleware/tracing.go]: not sure i remember touching tracing.go, but alright

File: response.go (hunk 13 above)
Comment (tiredkangaroo) [reply response.go:176]: we dont know which one the user wants so it doesn't really matter.

**Review by nikumar1206 — APPROVED** [review]

**Pull request merged**